File: docking/perl/peptide_docking.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Math::Trig;      # For trigonometric functions
use Math::Complex;  # For complex numbers (optional)

# ============================
# Define Peptides
# ============================

# Peptide 1: Simplified coordinates for 5 amino acids (e.g., Alanine)
my @peptide1 = (
    { name => 'Ala1', x => 0.0, y => 0.0, z => 0.0 },
    { name => 'Ala2', x => 1.5, y => 0.0, z => 0.0 },
    { name => 'Ala3', x => 3.0, y => 0.0, z => 0.0 },
    { name => 'Ala4', x => 4.5, y => 0.0, z => 0.0 },
    { name => 'Ala5', x => 6.0, y => 0.0, z => 0.0 },
);

# Peptide 2: Simplified coordinates for 5 amino acids (e.g., Glycine)
my @peptide2 = (
    { name => 'Gly1', x => 0.0, y => 0.0, z => 5.0 },
    { name => 'Gly2', x => 1.5, y => 0.0, z => 5.0 },
    { name => 'Gly3', x => 3.0, y => 0.0, z => 5.0 },
    { name => 'Gly4', x => 4.5, y => 0.0, z => 5.0 },
    { name => 'Gly5', x => 6.0, y => 0.0, z => 5.0 },
);

# ============================
# Docking Parameters
# ============================

my @rot_angles = (0, 45, 90, 135, 180, 225, 270, 315); # Degrees around Z-axis
my $translation_step = 1.0;  # Ångströms
my $max_translation = 5.0;   # Maximum translation in each direction (X and Y)
my $translation_range = (-$max_translation .. $max_translation); # Not directly usable in Perl, will handle in loops

# Output file for trajectory
my $trajectory_file = 'docking_trajectory.xyz';

# Open the trajectory file for writing
open(my $traj_fh, '>', $trajectory_file) or die "Cannot open $trajectory_file for writing: $!";

# ============================
# Docking Process
# ============================

my $best_score = 1e10;  # Initialize with a large number
my @best_pose = ();
my $best_rotation = 0;
my ($best_tx, $best_ty) = (0, 0);

# Initialize a frame counter for XYZ file
my $frame_counter = 0;

# Iterate over rotation angles
foreach my $angle_deg (@rot_angles) {
    my $angle_rad = deg2rad($angle_deg);
    my ($cos_a, $sin_a) = (cos($angle_rad), sin($angle_rad));

    # Iterate over translations in X direction
    for (my $tx = -$max_translation; $tx <= $max_translation; $tx += $translation_step) {
        # Iterate over translations in Y direction
        for (my $ty = -$max_translation; $ty <= $max_translation; $ty += $translation_step) {

            # Apply rotation and translation to Peptide 2
            my @docked_peptide2 = ();
            foreach my $aa (@peptide2) {
                # Rotate around Z-axis
                my $x_rot = $aa->{x} * $cos_a - $aa->{y} * $sin_a;
                my $y_rot = $aa->{x} * $sin_a + $aa->{y} * $cos_a;
                my $z_rot = $aa->{z};  # No change in Z for rotation around Z-axis

                # Translate
                my $x_new = $x_rot + $tx;
                my $y_new = $y_rot + $ty;
                my $z_new = $z_rot;

                push @docked_peptide2, { name => $aa->{name}, x => $x_new, y => $y_new, z => $z_new };
            }

            # Calculate score for this pose
            my $score = calculate_score(\@peptide1, \@docked_peptide2);

            # Update best score and pose if necessary
            if ($score < $best_score) {
                $best_score = $score;
                @best_pose = @docked_peptide2;
                $best_rotation = $angle_deg;
                ($best_tx, $best_ty) = ($tx, $ty);
            }

            # Log this pose to the trajectory file
            # Each frame in XYZ requires:
            # 1. Number of atoms
            # 2. Comment line
            # 3. Atom coordinates

            # Total number of atoms: 5 (Peptide1) + 5 (Docked Peptide2) = 10
            my $num_atoms = scalar(@peptide1) + scalar(@docked_peptide2);
            my $comment = "Frame $frame_counter: Rotation=$angle_deg°, Translation=($tx, $ty) Å, Score=$score";
            print $traj_fh "$num_atoms\n$comment\n";

            # Print Peptide1 atoms
            foreach my $aa1 (@peptide1) {
                printf $traj_fh "%-4s %8.3f %8.3f %8.3f\n", $aa1->{name}, $aa1->{x}, $aa1->{y}, $aa1->{z};
            }

            # Print Docked Peptide2 atoms
            foreach my $aa2 (@docked_peptide2) {
                printf $traj_fh "%-4s %8.3f %8.3f %8.3f\n", $aa2->{name}, $aa2->{x}, $aa2->{y}, $aa2->{z};
            }

            $frame_counter++;
        }
    }
}

# Close the trajectory file
close($traj_fh);
print "Trajectory saved to $trajectory_file\n";

# ============================
# Output Best Docking Pose
# ============================

print "Best Docking Score: $best_score Å\n";
print "Best Docking Rotation: $best_rotation degrees\n";
print "Best Docking Translation: X=$best_tx Å, Y=$best_ty Å\n";
print "\nPeptide 1 Coordinates:\n";
print_peptide(\@peptide1);
print "\nBest Docked Peptide 2 Coordinates:\n";
print_peptide(\@best_pose);

# ============================
# Subroutines
# ============================

# Calculate the docking score based on minimum distance between any pair of amino acids
sub calculate_score {
    my ($pep1_ref, $pep2_ref) = @_;
    my $min_dist = 1e10;

    foreach my $aa1 (@$pep1_ref) {
        foreach my $aa2 (@$pep2_ref) {
            my $dist = euclidean_distance($aa1->{x}, $aa1->{y}, $aa1->{z},
                                         $aa2->{x}, $aa2->{y}, $aa2->{z});
            if ($dist < $min_dist) {
                $min_dist = $dist;
            }
        }
    }
    return $min_dist;
}

# Calculate Euclidean distance between two points
sub euclidean_distance {
    my ($x1, $y1, $z1, $x2, $y2, $z2) = @_;
    return sqrt( ($x1 - $x2)**2 + ($y1 - $y2)**2 + ($z1 - $z2)**2 );
}

# Print peptide coordinates
sub print_peptide {
    my ($pep_ref) = @_;
    foreach my $aa (@$pep_ref) {
        printf "%-5s: (%.2f, %.2f, %.2f)\n", $aa->{name}, $aa->{x}, $aa->{y}, $aa->{z};
    }
}

File: docking/cpp/peptide_docking.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <string>
#include <limits>
#include <sstream> // Added to use std::ostringstream

// Constants
const double PI = 3.14159265358979323846;

// Structure to represent an Atom (Amino Acid)
struct Atom {
    std::string name;
    double x;
    double y;
    double z;

    Atom(std::string n, double xpos, double ypos, double zpos)
        : name(n), x(xpos), y(ypos), z(zpos) {}
};

// Structure to represent a Docking Pose
struct Pose {
    std::vector<Atom> docked_peptide2;
    double rotation_deg;
    double translation_x;
    double translation_y;
    double score;

    Pose() : rotation_deg(0.0), translation_x(0.0), translation_y(0.0), score(std::numeric_limits<double>::max()) {}
};

// Function to calculate Euclidean distance between two atoms
double calculate_distance(const Atom& a1, const Atom& a2) {
    double dx = a1.x - a2.x;
    double dy = a1.y - a2.y;
    double dz = a1.z - a2.z;
    return std::sqrt(dx * dx + dy * dy + dz * dz);
}

// Function to calculate docking score (minimum distance between any pair of amino acids)
double calculate_score(const std::vector<Atom>& peptide1, const std::vector<Atom>& docked_peptide2) {
    double min_dist = std::numeric_limits<double>::max();
    for (const auto& aa1 : peptide1) {
        for (const auto& aa2 : docked_peptide2) {
            double dist = calculate_distance(aa1, aa2);
            if (dist < min_dist) {
                min_dist = dist;
            }
        }
    }
    return min_dist;
}

// Function to apply rotation around Z-axis and translate
Atom rotate_atom(const Atom& aa, double cos_a, double sin_a, double tx, double ty) {
    double x_rot = aa.x * cos_a - aa.y * sin_a;
    double y_rot = aa.x * sin_a + aa.y * cos_a;
    double z_rot = aa.z; // No change in Z for rotation around Z-axis

    // Apply translation
    double x_new = x_rot + tx;
    double y_new = y_rot + ty;
    double z_new = z_rot;

    return Atom(aa.name, x_new, y_new, z_new);
}

// Function to print peptide coordinates to console
void print_peptide(const std::vector<Atom>& peptide) {
    for (const auto& aa : peptide) {
        std::cout << std::left << std::setw(5) << aa.name << ": ("
                  << std::fixed << std::setprecision(2)
                  << aa.x << ", " << aa.y << ", " << aa.z << ")\n";
    }
}

int main() {
    // ============================
    // Define Peptides
    // ============================

    // Peptide 1: Simplified coordinates for 5 amino acids (e.g., Alanine)
    std::vector<Atom> peptide1 = {
        Atom("Ala1", 0.0, 0.0, 0.0),
        Atom("Ala2", 1.5, 0.0, 0.0),
        Atom("Ala3", 3.0, 0.0, 0.0),
        Atom("Ala4", 4.5, 0.0, 0.0),
        Atom("Ala5", 6.0, 0.0, 0.0)
    };

    // Peptide 2: Simplified coordinates for 5 amino acids (e.g., Glycine)
    std::vector<Atom> peptide2 = {
        Atom("Gly1", 0.0, 0.0, 5.0),
        Atom("Gly2", 1.5, 0.0, 5.0),
        Atom("Gly3", 3.0, 0.0, 5.0),
        Atom("Gly4", 4.5, 0.0, 5.0),
        Atom("Gly5", 6.0, 0.0, 5.0)
    };

    // ============================
    // Docking Parameters
    // ============================

    std::vector<int> rot_angles = {0, 45, 90, 135, 180, 225, 270, 315}; // Degrees around Z-axis
    double translation_step = 1.0;  // Ångströms
    double max_translation = 5.0;    // Maximum translation in each direction (X and Y)

    // Output file for trajectory
    std::string trajectory_file = "docking_trajectory.xyz";

    // Open the trajectory file for writing
    std::ofstream traj_fh(trajectory_file);
    if (!traj_fh.is_open()) {
        std::cerr << "Cannot open " << trajectory_file << " for writing.\n";
        return 1;
    }

    // ============================
    // Docking Process
    // ============================

    Pose best_pose;
    int frame_counter = 0;

    for (const auto& angle_deg : rot_angles) {
        double angle_rad = angle_deg * PI / 180.0;
        double cos_a = std::cos(angle_rad);
        double sin_a = std::sin(angle_rad);

        for (double tx = -max_translation; tx <= max_translation; tx += translation_step) {
            for (double ty = -max_translation; ty <= max_translation; ty += translation_step) {

                // Apply rotation and translation to Peptide 2
                std::vector<Atom> docked_peptide2;
                for (const auto& aa : peptide2) {
                    Atom rotated_translated_aa = rotate_atom(aa, cos_a, sin_a, tx, ty);
                    docked_peptide2.push_back(rotated_translated_aa);
                }

                // Calculate score for this pose
                double score = calculate_score(peptide1, docked_peptide2);

                // Update best score and pose if necessary
                if (score < best_pose.score) {
                    best_pose.score = score;
                    best_pose.docked_peptide2 = docked_peptide2;
                    best_pose.rotation_deg = angle_deg;
                    best_pose.translation_x = tx;
                    best_pose.translation_y = ty;
                }

                // Log this pose to the trajectory file
                // Each frame in XYZ requires:
                // 1. Number of atoms
                // 2. Comment line
                // 3. Atom coordinates

                int num_atoms = peptide1.size() + docked_peptide2.size();
                std::ostringstream comment;
                comment << "Frame " << frame_counter << ": Rotation=" << angle_deg << "°, "
                        << "Translation=(" << tx << ", " << ty << ") Å, Score=" << score;

                traj_fh << num_atoms << "\n" << comment.str() << "\n";

                // Print Peptide1 atoms
                for (const auto& aa1 : peptide1) {
                    traj_fh << std::left << std::setw(5) << aa1.name << " "
                            << std::fixed << std::setprecision(3)
                            << std::setw(8) << aa1.x << " "
                            << std::setw(8) << aa1.y << " "
                            << std::setw(8) << aa1.z << "\n";
                }

                // Print Docked Peptide2 atoms
                for (const auto& aa2 : docked_peptide2) {
                    traj_fh << std::left << std::setw(5) << aa2.name << " "
                            << std::fixed << std::setprecision(3)
                            << std::setw(8) << aa2.x << " "
                            << std::setw(8) << aa2.y << " "
                            << std::setw(8) << aa2.z << "\n";
                }

                frame_counter++;
            }
        }
    }

    // Close the trajectory file
    traj_fh.close();
    std::cout << "Trajectory saved to " << trajectory_file << "\n";

    // ============================
    // Output Best Docking Pose
    // ============================

    std::cout << "Best Docking Score: " << best_pose.score << " Å\n";
    std::cout << "Best Docking Rotation: " << best_pose.rotation_deg << " degrees\n";
    std::cout << "Best Docking Translation: X=" << best_pose.translation_x << " Å, Y=" << best_pose.translation_y << " Å\n\n";

    std::cout << "Peptide 1 Coordinates:\n";
    print_peptide(peptide1);
    std::cout << "\nBest Docked Peptide 2 Coordinates:\n";
    print_peptide(best_pose.docked_peptide2);

    return 0;
}
